From 2d8e0b002870d8677b79256ea4fbf4a7624b9483 Mon Sep 17 00:00:00 2001
From: Bin Han <bin.b.han@intel.com>
Date: Mon, 20 Oct 2014 10:12:32 +0800
Subject: [PATCH 154/164] atomisp2: print out the code of FW_ERROR event and
 FW_WARNING event

BZ: 226908

For debug purpose, print out the code when FW_ERROR and FW_WARNING happen.

Change-Id: Idd3d406fc0653bbb56d297d1c03b637b65f4659a
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-8929
Signed-off-by: Bin Han <bin.b.han@intel.com>
Reviewed-on: https://android.intel.com:443/290504
Reviewed-on: https://icggerrit.ir.intel.com/29114
Approver: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../atomisp2/atomisp_driver/atomisp_compat_css20.c |   14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
index b8f69ce..4676d04 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
@@ -4554,6 +4554,20 @@ int atomisp_css_isr_thread(struct atomisp_device *isp,
 	struct atomisp_sub_device *asd = &isp->asd[0];
 
 	while (!atomisp_css_dequeue_event(&current_event)) {
+		if (current_event.event.type == IA_CSS_EVENT_TYPE_FW_ERROR) {
+			/* Received FW error signal, trigger WDT to recover */
+			dev_err(isp->dev, "%s: ISP reports FW_ERROR event, error code %d!!!!",
+			        __func__, current_event.event.fw_error);
+			atomisp_wdt_stop(isp, 0);
+			atomisp_wdt((unsigned long)isp);
+			return -EINVAL;
+		} else if (current_event.event.type == IA_CSS_EVENT_TYPE_FW_WARNING) {
+			dev_warn(isp->dev, "%s: ISP reports warning, code is %d, exp_id %d\n",
+			        __func__, current_event.event.fw_warning,
+			        current_event.event.exp_id);
+			continue;
+		}
+
 		asd = __get_atomisp_subdev(current_event.event.pipe,
 					isp, &stream_id);
 		if (!asd) {
-- 
1.7.9.5

